rasa/utils/tensorflow/layers.py

```diff
@@ -806,38 +801,32 @@ def _loss_margin(
         sim_neg_li: tf.Tensor,
         mask: Optional[tf.Tensor],
     ) -> tf.Tensor:
-        """Define max margin loss."""
+        """Define max margin loss.
 
-        # loss for maximizing similarity with correct action
-        loss = tf.maximum(0.0, self.mu_pos - tf.squeeze(sim_pos, axis=-1))
+        Can be interpreted as Triplet loss from https://arxiv.org/abs/1503.03832
+        """
+        # loss for maximizing similarity with correct label
+        # we have 4 negative losses, so this serves as 4 anchors in triplet loss
+        loss = 4 * tf.maximum(0.0, self.mu_pos - tf.squeeze(sim_pos, axis=-1))
 
-        # loss for minimizing similarity with `num_neg` incorrect actions
+        # loss for minimizing similarity with `num_neg` incorrect labels
         if self.use_max_sim_neg:
-            # minimize only maximum similarity over incorrect actions
+            # minimize only maximum similarity over incorrect labels
             max_sim_neg_il = tf.reduce_max(sim_neg_il, axis=-1)
             loss += tf.maximum(0.0, self.mu_neg + max_sim_neg_il)
         else:
-            # minimize all similarities with incorrect actions
+            # minimize all similarities with incorrect labels
             max_margin = tf.maximum(0.0, self.mu_neg + sim_neg_il)
             loss += tf.reduce_sum(max_margin, axis=-1)
```

Comment:
This should be `reduce_mean` to avoid extra negative terms as we discussed on slack.

Reply:
good point!